src/test/java/BaseTest.java

```diff
@@ -1,11 +1,181 @@
+    }
+
+    public void searchSong(String songTitleKeyword) {
+        WebElement searchSongInput = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='search']")));
+        searchSongInput.click();
+        searchSongInput.clear();
+        searchSongInput.sendKeys(songTitleKeyword);
+    }
+
+    public void clickViewAllSearchBtn() {
+        WebElement ViewAllSearchButton = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("#searchExcerptsWrapper > div > div > section.songs > h1 > button")));
+        ViewAllSearchButton.click();
+    }
+
+    public void clickFirstSongSearchResult() {
+        WebElement FirstSongSearchResult = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("#songResultsWrapper > div > div > div.item-container > table > tr")));
+        FirstSongSearchResult.click();
+    }
+
+    public void clickAddToBtn() {
+        WebElement AddToButton = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("#songResultsWrapper > header > div.song-list-controls > span > button.btn-add-to")));
+        AddToButton.click();
+    }
+
+    public void selectPlayList(){
+        WebElement selectedPlaylist = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id='songResultsWrapper']/header/div[3]/div/section[1]/ul/li[5]")));
+        selectedPlaylist.click();
+    }
+
+    public String getNotificationMessage() {
+        WebElement notificationAlert = driver.findElement(By.cssSelector("div.alertify-logs.top.right"));
+        return notificationAlert.getText();
+    }
+    public void songPlay() {
+        WebElement playNextButton = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id='mainFooter']/div[1]/i[2]")));
+        WebElement playButton = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id='mainFooter']/div[1]/span/span[2]")));
+        playNextButton.click();
+        playButton.click();
+    }
+
+    public Boolean isSongPlaying() {
+        WebElement soundBar = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id='mainFooter']/div[2]/div[2]/div/button[1]/div")));
+        return soundBar.isDisplayed();
+    }
+    // Helper functions for delete playlist
+        public void openPlayList(){
+        WebElement emptyPlaylist = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".playlist:nth-child(3)")));
+        emptyPlaylist.click();
+    }
+
+    public void deletePlaylistBtn(){
+        WebElement deletePlaylist = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".btn-delete-playlist")));
+        deletePlaylist.click();
+    }
+    public String getDeletedPlaylistMsg()    {
+        WebElement deletedNotificationMsg = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("div.success.show")));
+        return deletedNotificationMsg.getText();
+    }
```

Comment:
Page Object Model (POM): recommended to associate the methods with the corresponding page class where they are relevant. In the POM, each page or section of a web application is represented by a separate class. These classes encapsulate the elements and actions specific to that page or section.

Move these methods to the corresponding class according to the POM.

Reply:
1. **BasePage and BaseTest** 

**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

![image](https://user-images.githubusercontent.com/104761482/233972627-2d7ffdc7-6a2f-4e9f-8a3a-a1b4a7a47bc4.png)


**BasePage**  — class serves as a foundation or parent class for individual page classes. It contains common methods and functionalities that are shared across multiple pages in an application.

_Example without a Base Page Class_

![image](https://user-images.githubusercontent.com/104761482/236891493-7f7013c7-fde2-4bf9-9084-3aa7483df8e3.png)


2. **Test in Page Object Model (POM)**

![Clean_Test](https://user-images.githubusercontent.com/104761482/236863813-f30d7bfc-e89d-4419-bdea-4b1517155e39.png)

3.  **Page Object Model and Page Factory**

![POM_vs_Factory_1](https://user-images.githubusercontent.com/104761482/236864030-34b89652-460d-4440-aa91-894332e90654.png)

![POM_vs_Factory_2](https://user-images.githubusercontent.com/104761482/236864098-d3ffe636-7ba9-484e-a72f-8b03ff024659.png)